OutOfSchool/OutOfSchool.DataAccess/Repository/ChangesLogRepository.cs

```diff
@@ -0,0 +1,118 @@
+
+namespace OutOfSchool.Services.Repository
+{
+    public class ChangesLogRepository : EntityRepository<ChangesLog>, IChangesLogRepository
+    {
+        public ChangesLogRepository(OutOfSchoolDbContext dbContext)
+            : base(dbContext)
+        {
+        }
+
+        public ICollection<ChangesLog> AddChangesLogToDbContext<TEntity>(
+            TEntity entity,
+            string userId,
+            IEnumerable<string> trackedProperties,
+            Func<Type, object, string> valueProjector)
+            where TEntity : class, IKeyedEntity, new()
+        {
+            var result = new List<ChangesLog>();
+            var entry = dbContext.Entry(entity);
+
+            var entityType = typeof(TEntity).Name;
+            var (entityIdGuid, entityIdLong) = GetEntityId(entry);
+            var changedValues = GetChangedValues(entry, trackedProperties, valueProjector);
+            foreach (var (propertyName, oldValue, newValue) in changedValues)
+            {
+                result.Add(CreateChangesLogRecord(
+                    entityType,
+                    propertyName,
+                    entityIdGuid,
+                    entityIdLong,
+                    oldValue,
+                    newValue,
+                    userId));
+            }
+
+            if (result.Count > 0)
+            {
+                dbContext.AddRange(result);
+            }
+
+            return result;
+        }
+
+        // TODO: logging of the Institution changes is yet to be configured
+        private IEnumerable<(string PropertyName, string OldValue, string NewValue)> GetChangedValues(
+            EntityEntry entityEntry,
+            IEnumerable<string> trackedProperties,
+            Func<Type, object, string> valueProjector)
+        {
+            var properties = entityEntry.Properties
+                .Where(p => p.IsModified
+                    && trackedProperties.Contains(p.Metadata.PropertyInfo.Name))
+                .Select(x => (x.Metadata.PropertyInfo.Name,
+                    valueProjector(x.Metadata.ClrType, x.OriginalValue),
+                    valueProjector(x.Metadata.ClrType, x.CurrentValue)));
+
+            var references = entityEntry.References
+                .Where(x => trackedProperties.Contains(x.Metadata.PropertyInfo.Name)
+                    && x.TargetEntry?.State == EntityState.Modified)
+                .Select(x => (x.Metadata.PropertyInfo.Name,
+                    valueProjector(x.TargetEntry.Metadata.ClrType, x.TargetEntry.OriginalValues.ToObject()),
+                    valueProjector(x.TargetEntry.Metadata.ClrType, x.TargetEntry.CurrentValues.ToObject())));
+
+            return properties.Concat(references);
+        }
+
+        private (Guid? entityIdGuid, long? entityIdLong) GetEntityId(EntityEntry entityEntry)
+        {
+            Guid? entityIdGuid = null;
+            long? entityIdLong = null;
+
+            var idProperty = entityEntry.Property("Id");
```

Comment:
I think it'd be better to get `string idColumnName` as parameter to expand usage.
And also just a comment I suppose you define which type id is? And then you add 2 ids as record in table, one of which is null. So it could be used with `string Id` and one more column as `IdType` (using enum). But I guess it adds a little bit more complexity, so just a comment. Maybe somebody can offer something.
Or if I don't understand, I'm sorry.

Reply:
Agree with that, that's why I've said 'just a comment for considering' :)